Add guidelines around being a repo maintainer (#12)
RMI-PACTA/practices: maintainer.md

# What Does It Mean to Maintain a Repository?

A solid ecosystem of maintainers can help lead to a well-functioning software environment. Maintainers are responsible for the health of a repository, which includes its functionality, documentation, and communication. Maintainers are also responsible for the health of the community around the repository, which includes the responsiveness of the maintainers, the clarity of the repository’s purpose, and the accessibility of the repository to new users.

> Below are some of our heuristics for what it means to be a maintainer of a repository. These are not necessarily exhaustive, but should give an idea of the expectations of a maintainer.

Reviewer: not sure "heuristics" is the appropriate word here... sorry, don't have a suggestion popping in my mind at the moment

Reviewer 2: guidelines?

## Communication

> Maintainers are responsible for clear communication regarding the state of the repository and its intended use, including any expected changes in usage. When possible, this may include clear documentation of its API.

Reviewer 2: ```suggestion
Maintainers are responsible for clear communication regarding the state of the repository and its intended use, including any expected changes in usage. This includes clear documentation of its interface (see Documentation, below).
```

## Updates and Fixes

> They must update and fix the repository in a timely manner, which involves bug fixes, reviewing issues, and updating workflows/ actions.

Reviewer 2: Maybe "They commit to addressing issues with this repository in a timely manner."?

## Documentation

> They must provide good documentation to explain the expected functionality and outputs of the repository.

Reviewer 2: Would like to expand with more definition of what "good" documentation consists of

Reviewer 2: Possibly a topic for a separate discussion, but documentation should include

* Purpose of repo
* Overall structure (of repo)
* How to use (main use case, additional usecases)
  * Invoking workflow (if applicable)
  * functions of interest
* How to contribute (development usecase)

Author: Resolved by a new issue/ practice specific to documentation.

## Decision Making

> Maintainers are the decision-makers for structural changes or contentious features within the repository.

Reviewer 2: With guidance/consultation from interested parties?

Author: I hope this would be heavily implied? But I would rather keep the "decision maker" definition clear and unambiguous...

Author: But yes, absolutely agree with the sentiment. If you feel strongly that we need to explicitly state it then I will add this.

## Awareness of Use and Dependencies
Maintainers should be particularly cautious about changes to repositories that are dependencies of larger workflows to avoid breaking changes.

## Communication of Changes
Any substantial change that may affect other parts of the workflow should be communicated, preferably in a tech review call or similar forum, to discuss the potential impacts and necessary adjustments.

> ## Collaborative Responsibility

Reviewer 3: NIT: does one code maintainer per repo also means we don't want to have two people assigned to bigger and more repos? Or are we focusing on maintainers here because we want clear responsibilities and will still define 'main contributor' or something like that elsewhere?

> Ideally, there will only be a one maintainer per repository. In some cases, that maintainer may decide that certain files (for example, the `Dockerfile`) are best maintained by someone else. In that case, a file-specific maintainer may be specified. The lead maintainer should be responsible for ensuring that the co-maintainer is aware of the changes and is able to make them. If more than a handful of files are managed by a co-maintainer, it is a good indication that the repository should be refactored and split into multiple repositories.

Reviewer: ```suggestion
Ideally, there will only be one maintainer per repository. In some cases, that maintainer may decide that certain files (for example, the `Dockerfile`) are best maintained by someone else. In that case, a file-specific maintainer may be specified. The lead maintainer should be responsible for ensuring that the co-maintainer is aware of the changes and is able to make them. If more than a handful of files are managed by a co-maintainer, it is a good indication that the repository should be refactored and split into multiple repositories.
```

## Practical Implications
Maintainers of a repository will be defined using the `.github/CODEOWNERS` file of that repository. File-specific maintainers can also be defined in this manner. See [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) for more info. 

> 

Author: I think that depends on #6 to be honest. I will extract that comment into an issue that can likely be merged after we discuss #6?

> 

Author: Also directly relates to #18

> 

Author: I would say yes crucial to document here, but we are not able to yet. 

But we will soon^TM

> 

Author: Resolving to look into this here: https://github.com/RMI-PACTA/practices/issues/18#issuecomment-1934143490
